### PR TITLE
refactor(core-transaction-pool): expire transactions that don't have …

### PR DESCRIPTION
**Pull request by vasild**
…expiration

If a transaction does not have an expiration value set by the user, then
we would keep it for a long time in the pool and also it could possibly
be re-sent by somebody (not the author) long time after it was created,
if it has not been forged yet.

Change this to expire after 6h even transactions that don't have an
user-set expiration. 6h considered since the transaction creation time.

## What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [x] Refactor
- [ ] Performance
- [ ] Tests
- [ ] Build
- [ ] Documentation
- [ ] Code style update
- [ ] Continuous Integration
- [ ] Other, please describe:

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Does this PR release a new version?

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `develop` branch, _not_ the `master` branch
- [ ] All tests are passing
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

### CODE REVIEW

File: packages/core-transaction-pool/src/memory.ts [hunk 1]
```diff
@@ -250,4 +275,43 @@ export class Memory {
...
+    /**
+     * Calculate the expiration height of a transaction.
+     * An expiration height H means that the transaction cannot be included in block at height
+     * H or any higher block.
+     * If the user did not specify an expiration height when creating the transaction then
+     * we calculate one from the timestamp of the transaction creation and the configured
+     * maximum transaction age.
+     * @return number expiration height or null if the transaction does not expire
+     */
+    private calcTxExpiration(
```
Comment (supaiku0) [inline packages/core-transaction-pool/src/memory.ts]: `calculateTransactionExpiration`

File: packages/core-transaction-pool/src/memory.ts [hunk 2]
```diff
@@ -250,4 +275,43 @@ export class Memory {
...
+        if (transaction.data.expiration > 0) {
+            return transaction.data.expiration;
+        }
+
+        // Since the user did not specify an expiration we set one by calculating
+        // approximately the height of the chain as of the time the transaction was
+        // created and adding maxTransactionAge to that.
+
+        // Both now and transaction.data.timestamp use [number of seconds since the genesis block].
+        const txCreatedSecondsAgo: number = context.now - transaction.data.timestamp;
```
Comment (supaiku0) [inline packages/core-transaction-pool/src/memory.ts]: Please remove the `tx` prefix.

File: packages/core-transaction-pool/src/memory.ts [hunk 3]
```diff
@@ -250,4 +275,43 @@ export class Memory {
...
+
+        const txCreatedBlocksAgo: number = Math.floor(txCreatedSecondsAgo / context.blockTime);
```
Comment (supaiku0) [inline packages/core-transaction-pool/src/memory.ts]: Same.

File: packages/core-transaction-pool/src/memory.ts [hunk 4]
```diff
@@ -250,4 +275,43 @@ export class Memory {
...
+
+        const txCreatedAtHeight: number = context.currentHeight - txCreatedBlocksAgo;
```
Comment (supaiku0) [inline packages/core-transaction-pool/src/memory.ts]: Same.

**Comment by alessiodf** [comment]
>... it could possibly be re-sent by somebody (not the author) long time after it was created, if it has not been forged yet.

I don't see how this PR mitigates that risk. Somebody who is not the author who wants to do that could easily add an arbitrary `expiration: someHeightInTheFuture` property to the transaction and broadcast it since the `expiration` value is not signed so can be manipulated and still pass validation and verification.

File: packages/core-transaction-pool/src/memory.ts (hunk 1 above)
Comment (vasild) [reply packages/core-transaction-pool/src/memory.ts]: Done in 7a9b97211.

File: packages/core-transaction-pool/src/memory.ts (hunk 2 above)
Comment (vasild) [reply packages/core-transaction-pool/src/memory.ts]: Done in 7a9b97211.

File: packages/core-transaction-pool/src/memory.ts (hunk 3 above)
Comment (vasild) [reply packages/core-transaction-pool/src/memory.ts]: Done in 7a9b97211.

File: packages/core-transaction-pool/src/memory.ts (hunk 4 above)
Comment (vasild) [reply packages/core-transaction-pool/src/memory.ts]: Done in 7a9b97211.

**Comment by vasild** [comment]
> > ... it could possibly be re-sent by somebody (not the author) long time after it was created, if it has not been forged yet.
> 
> I don't see how this PR mitigates that risk. Somebody who is not the author who wants to do that could easily add an arbitrary `expiration: someHeightInTheFuture` property to the transaction and broadcast it since the `expiration` value is not signed so can be manipulated and still pass validation and verification.

Right, good catch! I assumed that the `expiration` field is signed, but that is the case only with v2 transactions. I changed the behavior to ignore the expiration field for v1 transactions (ie proceed as if it is 0) in 78d658499.

Thanks!

**Comment by vasild** [comment]
> Please also add a test for it.

See the updated test at:

https://github.com/ArkEcosystem/core/pull/2672/files#diff-b0a172ced81203aed79dde98f43fc8e0R233

It creates a few transactions with version=1 and version=2 with different expiration values.

**Pull request merged**